### PR TITLE
Deprecate existing WorkerBinder.bind methods and provide recommended bindTo methods with RibDispatchers.Default

### PR DESCRIPTION
**Pull request by FranAguilera**
**Description**:

Since https://github.com/uber/RIBs/pull/553 we added support to specify on which CoroutineDispatcher are presidio Worker will be bound. RibDispatchers.Unconfined was added as a default parameter in order to keep existing behavior and not be a breaking change. 

Still, binding by default on the current caller thread within interactor means that by default we are binding workers on main UI thread leading to potential jank ANR etc just for binding those workers. 

This is specially problematic when the list of workers being bound keeps increasing (e.g. hypothetically if we have 100 workers and each bind takes at least 50 ms this will result on the UI thread being blocked for at least 5000ms)

Proposal is to deprecate existing methods and provide new methods that by default will bind on **Ribdispatchers.DEFAULT** (similarly to **RibCoroutineWorker**). Still if there is a need for a Worker to change the thread where it will be bound we can rely on **coroutineContext** property to be overriden as following (default value is **EmptyCoroutineContext**):

```
  class MyUiWorker:com.uber.rib.core.Worker{
    override val coroutineContext: CoroutineContext = RibDispatchers.Main
  }
```

Demo call:

`worker.bindTo(interactor)
`

NOTE: For calling on existing java files -> `WorkerBinderKt.bindTo(ribWorker, this));`

**Related issue(s)**:

https://github.com/uber/RIBs/issues/595

Verified via Unit tests

### CODE REVIEW

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 1]
```diff
@@ -59,6 +59,46 @@ public object WorkerBinder {
     this.workerBinderListenerWeakRef = WeakReference<WorkerBinderListener>(workerBinderListener)
   }
 
+  /**
+   * Bind a worker (ie. a manager or any other class that needs an interactor's lifecycle) to an
+   * interactor's lifecycle events
+   *
+   * IMPORTANT: Binding will happen on [RibDispatchers.DEFAULT] unless the Worker explicitly
```
Comment (psteiger) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:66]: Non-caps `RibDispatchers.Default`

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 2]
```diff
@@ -59,6 +59,46 @@ public object WorkerBinder {
...
+  ): WorkerUnbinder = bind(interactor, worker, RibDispatchers.Default)
+
+  /**
+   * Bind a worker (ie. a manager or any other class that needs an interactor's lifecycle) to an
+   * interactor's lifecycle events
+   *
+   * IMPORTANT: Binding will happen on [RibDispatchers.DEFAULT] unless the Worker explicitly
+   * overrides [coroutineContext] to other than [EmptyCoroutineContext]
+   *
+   * Inject this class into your interactor and call this method on any
```
Comment (psteiger) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:88]: On any... ?

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 3]
```diff
@@ -59,6 +59,46 @@ public object WorkerBinder {
...
   }
 
+  /**
+   * Bind a worker (ie. a manager or any other class that needs an interactor's lifecycle) to an
+   * interactor's lifecycle events
+   *
+   * IMPORTANT: Binding will happen on [RibDispatchers.DEFAULT] unless the Worker explicitly
+   * overrides [coroutineContext] to other than [EmptyCoroutineContext]
+   *
+   * Inject this class into your interactor and call this method on any
```
Comment (psteiger) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:69]: On any... ?

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 4]
```diff
@@ -59,6 +59,46 @@ public object WorkerBinder {
...
+   *
+   * @param interactor The interactor that provides the lifecycle.
+   * @param worker The class that wants to be informed when to start and stop doing work.
+   * @return [WorkerUnbinder] to unbind [Worker]'s lifecycle.
+   */
+  @JvmStatic
+  public fun bindToInteractor(
+    interactor: Interactor<*, *>,
+    worker: Worker,
```
Comment (psteiger) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:78]: Do you feel like we should provide an optional `dispatcher` parameter like we had before, but defaulting to `RibDispatchers.Default` instead?

```
  @JvmStatic @JvmOverloads
  public fun bindToInteractor(
    interactor: Interactor<*, *>,
    worker: Worker,
    dispatcher: CoroutineDispatcher = RibDispatchers.Default,
  ): WorkerUnbinder = bind(interactor, worker, dispatcher)
```

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 5]
```diff
@@ -107,6 +163,42 @@ public object WorkerBinder {
...
+   * overrides [coroutineContext] to other than [EmptyCoroutineContext]
+   *
+   * @param presenter The presenter that provides the lifecycle.
+   * @param worker The class that wants to be informed when to start and stop doing work.
+   * @return [WorkerUnbinder] to unbind [Worker]'s lifecycle.
+   */
+  @JvmStatic
+  public fun bindToPresenter(
+    presenter: Presenter,
+    worker: Worker,
```
Comment (psteiger) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:180]: Default dispatcher parameter?

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 6]
```diff
@@ -107,6 +163,42 @@ public object WorkerBinder {
...
+   * overrides [coroutineContext] to other than [EmptyCoroutineContext]
+   *
+   * @param presenter The presenter that provides the lifecycle.
+   * @param workers A list of classes that want to be informed when to start and stop doing work.
+   * @return [WorkerUnbinder] to unbind [Worker]'s lifecycle.
+   */
+  @JvmStatic
+  public fun bindToPresenter(
+    presenter: Presenter,
+    workers: List<Worker>,
```
Comment (psteiger) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:197]: Default dispatcher parameter?

**Review by psteiger — CHANGES_REQUESTED** [review]

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt (hunk 1 above)
Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:66]: good catch, will update

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt (hunk 3 above)
Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:69]: uops, yeah I copy pasted from the previous javadoc but that was clearly off. Will update in a bit

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt (hunk 4 above)
Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:78]: I guess we could keep it but at the same time I'm not too sold on this given the worker can specify its own dispatcher to override the one passed at WorkerBinder. 

If there is a need to specify another dispatcher probably is best to define within Worker level for clarification and to prevent unwanted changes outside the worker definition

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt (hunk 6 above)
Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:197]: discussed above

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt (hunk 4 above)
Comment (psteiger) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:78]: I'm fine with that

**Review by psteiger — APPROVED** [review]

**Comment by FranAguilera** [comment]
Sorry about the back and forth @psteiger, decided to take advantage of this and provide a migration API for all current WorkerBinder.bind calls. Will share internal doc offline too

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 7]
```diff
@@ -243,6 +389,19 @@ public object WorkerBinder {
       }
     }
   }
+
+  private fun getDispatcherFromMigration(
```
Comment (tyvsmith) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:393]: The set of APIs around the migration feels like we're leaking to much internal business logic into the public APis of Ribs. We should generalize this more. Perhaps just one override where we can set a static overriding dispatcher.

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 8]
```diff
@@ -95,21 +168,74 @@ public object WorkerBinder {
...
+   *
+   * IMPORTANT: Binding will happen on [RibDispatchers.Default] unless the Worker explicitly
+   * overrides [coroutineContext] to other than [EmptyCoroutineContext]
+   *
+   * @param presenter The presenter that provides the lifecycle.
+   * @param workers A list of classes that want to be informed when to start and stop doing work.
+   * @return [WorkerUnbinder] to unbind [Worker]'s lifecycle.
+   */
+  @JvmStatic
+  public fun bindToPresenter(
```
Comment (tyvsmith) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:227]: This looks like new functionality added that isn't needed for these changes (unless I missed it's usage) where we support binding to a list.

1. Should this be a separate PR if we want a new puiblic API.
2. Rather in this diff or on the next, probably rename to bindToPresenters (plural)

Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:227]: We've been providing these APIs (_**WorkerBinder.bind(interactor, listOfWorkers)/ WorkerBinder.bind(presenter, listOfWorkers**_) for a while. 

Part of the goal of this PR is to deprecate existing WorkerBinder.bind methods given that the methods with 2 parameter call takes a default RibDispatchers.Unconfined which often times results in binding on main thread. 

Was thinking that adding new methods (bindToInteractor, bindToPresenter) that are opinionated on using to RibDispatchers.Default to prevent further additions of WorkerBinder.bind on main thread (we can easily rely on static analysis to force everyone to use the new API when new WorkerBinder additions). 

Also another important caveat, if there is still a need for a legacy worker to be bound on the main thread this can be achieved by overriding Worker's coroutineContext property (this will ignore the default Dispatchers specified on WorkerBinder call)

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt (hunk 7 above)
Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:393]: I agree but also I'm a bit concerned that providing an easy static dispatcher override could lead to unwanted dispatchers change at any point by any consumers.

Let me think about this a bit more

Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt]: @tyvsmith Decided to keep this migration in internal monorepo. This PR will just focus on deprecating existing API and add equivalent new methods for the default binding happening off the main thread

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 9]
```diff
@@ -59,17 +59,64 @@ public object WorkerBinder {
...
+   * interactor's lifecycle events
+   *
+   * IMPORTANT: Binding will happen on [RibDispatchers.Default] unless the Worker explicitly
+   * overrides [coroutineContext] to other than [EmptyCoroutineContext]
+   *
+   * @param interactor The interactor that provides the lifecycle.
+   * @return [WorkerUnbinder] to unbind [Worker]'s lifecycle.
+   */
+  @JvmStatic
+  public fun Worker.bindTo(
```
Comment (psteiger) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:73]: +1 to extension function. 

But since we're providing those new APIs as public extension function, I'd just remove them from inside the `WorkerBinder` object.

Consider it a nit.

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt [hunk 10]
```diff
@@ -108,8 +163,48 @@ public object WorkerBinder {
...
+   * presenter's lifecycle events.
+   *
+   * IMPORTANT: Binding will happen on [RibDispatchers.Default] unless the Worker explicitly
+   * overrides [coroutineContext] to other than [EmptyCoroutineContext]
+   *
+   * @param presenter The presenter that provides the lifecycle.
+   * @return [WorkerUnbinder] to unbind [Worker]'s lifecycle.
+   */
+  @JvmStatic
+  public fun Worker.bindTo(
```
Comment (psteiger) [inline android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:176]: Same

File: android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt (hunk 9 above)
Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt:73]: good catch! will be moving this outside WorkerBinder object in a sec

Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt]: @psteiger Decided to move all existing extensions to its own file so we can call the new extension functions on legacy java code

Comment (FranAguilera) [reply android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt]: Please ignore this. Reverted back to current WorkerBinder

**Comment by FranAguilera** [comment]
> I don't have any conceptual objections. Exposing new API as public top-level function instead of nested inside `WorkerBinder` object is nice. And deprecating the previous API instead of changing it keeps retro-compat.
> 
> With that said, maybe we should consider deprecating `Worker` in favor of `RibCoroutineWorker`. I'm not sure it's worth to keep evolving Worker API. But if it's being done as a smoother transition pathway, I'm ok with it.

I completely agree with eventually fully deprecating this old worker. The only missing piece for RibCoroutineWorker is monitoring option for binding duration (but given that by default is bound on RibDispatchers.Default. there is no much of a concern). Also we can validate a few of this RibCoroutineWorkers in prod and if no issues found we can completely mark worker as deprecated

**Comment by FranAguilera** [comment]
> > I don't have any conceptual objections. Exposing new API as public top-level function instead of nested inside `WorkerBinder` object is nice. And deprecating the previous API instead of changing it keeps retro-compat.
> > With that said, maybe we should consider deprecating `Worker` in favor of `RibCoroutineWorker`. I'm not sure it's worth to keep evolving Worker API. But if it's being done as a smoother transition pathway, I'm ok with it.
> 
> I completely agree with eventually fully deprecating this old worker. The only missing piece for RibCoroutineWorker is monitoring option for binding duration (but given that by default is bound on RibDispatchers.Default. there is no much of a concern). Also we can validate a few of this RibCoroutineWorkers in prod and if no issues found we can completely mark worker as deprecated

Actually, after thinking about a bit more I don't think we should keep expanding this WorkerBinder API given is going to be deprecated soon in favor of RibCoroutineWorker.

I was thinking though to add a migration mechanism to change the WorkerBinderDispatcher to other than Unconfined. In that manner we can migrate internally all worker binding outside Ui thread in a safe manner

**Comment by FranAguilera** [comment]
Will be putting a new PR for migration API options
